python/ray/data/datasource/binary_datasource.py

```diff
@@ -16,8 +17,24 @@ class BinaryDatasource(FileBasedDatasource):
     """
 
     def _read_file(self, f: "pyarrow.NativeFile", path: str, **reader_args):
+        import pyarrow
+        from pyarrow.fs import HadoopFileSystem
+
         include_paths = reader_args.pop("include_paths", False)
-        data = f.readall()
+        if reader_args.get("compression") == "snappy":
+            import snappy
+
+            filesystem = reader_args.get("filesystem", None)
+            rawbytes = BytesIO()
+
+            if isinstance(filesystem, pyarrow.fs.HadoopFileSystem):
+                snappy.hadoop_snappy.stream_decompress(src=f, dst=rawbytes)
+            else:
+                snappy.stream_decompress(src=f, dst=rawbytes)
```

Comment:
I double-checked [this API](https://github.com/andrix/python-snappy/blob/c4411aedb6b719f3aa27f9ae80aa7f8c1bd26ae1/src/snappy/snappy.py#L298-L323) to see if there was anything else that we wanted to expose to the user via something like `ray.data.read_binary(..., decompress_args={...})`, but I think that we're good. E.g., snappy says that the default block size should work for nearly all payloads.